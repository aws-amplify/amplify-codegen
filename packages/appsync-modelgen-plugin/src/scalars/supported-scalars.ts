//  Used only in  tests. Directive definition will be passed as part of the configuration when modelgen is run using CLI
export const scalars = [
  'ID',
  'String',
  'Int',
  'Float',
  'Boolean',
  'AWSDate',
  'AWSDateTime',
  'AWSTime',
  'AWSTimestamp',
  'AWSEmail',
  'AWSJSON',
  'AWSURL',
  'AWSPhone',
  'AWSIPAddress',
]
  .map(typeName => `scalar ${typeName}`)
  .join();
