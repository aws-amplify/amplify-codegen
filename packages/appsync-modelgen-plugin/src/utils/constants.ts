export const TransformerV2DirectiveName = {
  HAS_ONE: 'hasOne',
  HAS_MANY: 'hasMany',
  BELONGS_TO: 'belongsTo',
  MODEL: 'model',
  AUTH: 'auth',
  PRIMARY_KEY: 'primaryKey',
  INDEX: 'index',
  DEFAULT: 'default',
  SEARCHABLE: 'searchable',
};
export const DEFAULT_HASH_KEY_FIELD = 'id';
export const DEFAULT_CREATED_TIME = 'createdAt';
export const DEFAULT_UPDATED_TIME = 'updatedAt';
