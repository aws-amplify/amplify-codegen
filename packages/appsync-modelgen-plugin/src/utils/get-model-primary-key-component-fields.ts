import { CodeGenField, CodeGenModel } from "../visitors/appsync-visitor";

export function getModelPrimaryKeyComponentFields(model: CodeGenModel): CodeGenField[] {
  const primaryKeyField = model.fields.find(field => field.primaryKeyInfo)!;
  const { sortKeyFields } = primaryKeyField.primaryKeyInfo!;
  return [ primaryKeyField, ...sortKeyFields ];
}