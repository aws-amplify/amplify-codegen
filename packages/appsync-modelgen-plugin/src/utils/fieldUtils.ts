import { CodeGenDirective, CodeGenField, CodeGenModel } from '../visitors/appsync-visitor';
import { TransformerV2DirectiveName } from './constants';

export function addFieldToModel(model: CodeGenModel, field: CodeGenField): void {
  const existingField = model.fields.find(f => f.name === field.name);
  if (!existingField) {
    model.fields.push(field);
  }
}

export function removeFieldFromModel(model: CodeGenModel, fieldName: string): void {
  model.fields = model.fields.filter(field => field.name !== fieldName);
}

export const getDirective = (fieldOrModel: CodeGenField | CodeGenModel) => (directiveName: string): CodeGenDirective | undefined =>
  fieldOrModel.directives.find(d => d.name === directiveName);

// Function matching to GraphQL transformer so that the auto-generated field
export function toCamelCase(words: string[]): string {
  const formatted = words.map((w, i) => (i === 0 ? w.charAt(0).toLowerCase() + w.slice(1) : w.charAt(0).toUpperCase() + w.slice(1)));
  return formatted.join('');
}

export function getOtherSideBelongsToField(type: string, otherSideModel: CodeGenModel): CodeGenField | undefined {
  return otherSideModel.fields.filter(f => f.type === type).find(f => f.directives.find(d => d.name === TransformerV2DirectiveName.BELONGS_TO));
}

/**
 * Given a model, it returns the primary and sort key fields if present, an empty list otherwise.
 * @param model Codegen Model object
 * @returns Array of primary and sort key codegen fields if present or an empty list
 */
export function getModelPrimaryKeyComponentFields(model: CodeGenModel): CodeGenField[] {
  const primaryKeyField = model.fields.find(field => field.primaryKeyInfo);
  const keyFields: CodeGenField[] = [];
  if (primaryKeyField) {
    keyFields.push(primaryKeyField);
    if ( primaryKeyField?.primaryKeyInfo?.sortKeyFields ) {
      keyFields.push(...primaryKeyField.primaryKeyInfo.sortKeyFields);
    };
  }
  return keyFields;
}