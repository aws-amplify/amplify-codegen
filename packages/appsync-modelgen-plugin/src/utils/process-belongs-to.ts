import { CodeGenDirective, CodeGenField, CodeGenModel, CodeGenModelMap } from '../visitors/appsync-visitor';
import { getModelPrimaryKeyComponentFields } from './get-model-primary-key-component-fields';
import {
  CodeGenConnectionType,
  CodeGenFieldConnection,
  flattenFieldDirectives,
  makeConnectionAttributeName,
} from './process-connections';
import { getConnectedFieldV2 } from './process-connections-v2';


export function processBelongsToConnection(
  field: CodeGenField,
  model: CodeGenModel,
  modelMap: CodeGenModelMap,
  connectionDirective: CodeGenDirective,
  useFieldNameForPrimaryKeyConnectionField: boolean = false,
): CodeGenFieldConnection | undefined {
  const otherSide = modelMap[field.type];
  const otherSideField = getConnectedFieldV2(field, model, otherSide, connectionDirective.name);
  const connectionFields = connectionDirective.arguments.fields || [];

  if (connectionFields.length > 1) {
    // Todo: Move to a common function and update the error message
    throw new Error('DataStore only support one key in field');
  }

  if (field.isList) {
    throw new Error(
      `A list field does not support the 'belongsTo' relation`
    );
  }

  let validOtherSideField = false;
  otherSideField.directives.forEach(dir => {
    if (dir.name === 'hasOne' || dir.name === 'hasMany') {
      validOtherSideField = true;
    }
  });

  if (!validOtherSideField) {
    throw new Error(
      `A 'belongsTo' field should match to a corresponding 'hasMany' or 'hasOne' field`
    );
  }
  // if a type is connected using name, then amplify-graphql-relational-transformer adds a field to
  //  track the connection and that field is not part of the selection set
  // but if the field are connected using fields argument in connection directive
  // we are reusing the field and it should be preserved in selection set
  const otherSideHasMany = otherSideField.isList;
  const isConnectingFieldAutoCreated = false;

  // New metada type introduced by custom PK v2 support
  const targetNames = connectionFields.length > 0
    ? [ ...connectionFields ]
    : (otherSideHasMany
      ? [makeConnectionAttributeName(otherSide.name, otherSideField.name)]
      : (useFieldNameForPrimaryKeyConnectionField
        ? getModelPrimaryKeyComponentFields(otherSide).map(componentField => makeConnectionAttributeName(model.name, field.name, componentField.name))
        : [makeConnectionAttributeName(model.name, field.name)]));

  return {
    kind: CodeGenConnectionType.BELONGS_TO,
    connectedModel: otherSide,
    isConnectingFieldAutoCreated,
    targetName: targetNames[0],
    targetNames,
  };
}

export function getBelongsToConnectedField(field: CodeGenField, model: CodeGenModel, connectedModel: CodeGenModel): CodeGenField | undefined {
  let otherSideDirectives = flattenFieldDirectives(connectedModel).filter(dir => {
    const connectedField = connectedModel.fields.find(connField => { return connField.name === dir.fieldName; });
    const fieldType = connectedField?.type;
    return ((dir.name === 'hasOne' && !connectedField?.isList) || (dir.name === 'hasMany' && connectedField?.isList)) && model.name === fieldType;
  });

  if (otherSideDirectives?.length === 1) {
    return connectedModel.fields.find(connField => { return connField.name === otherSideDirectives[0].fieldName; });
  }
}

