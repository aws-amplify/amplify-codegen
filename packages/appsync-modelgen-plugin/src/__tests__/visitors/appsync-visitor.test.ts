import { buildSchema, parse, visit } from 'graphql';
import { directives, scalars } from '../../scalars/supported-directives';
import { CodeGenConnectionType, CodeGenFieldConnectionBelongsTo, CodeGenFieldConnectionHasMany } from '../../utils/process-connections';
import { AppSyncModelVisitor, CodeGenField, CodeGenGenerateEnum } from '../../visitors/appsync-visitor';

const buildSchemaWithDirectives = (schema: String) => {
  return buildSchema([schema, directives, scalars].join('\n'));
};

const createAndGenerateVisitor = (schema: string) => {
  const ast = parse(schema);
  const builtSchema = buildSchemaWithDirectives(schema);
  const visitor = new AppSyncModelVisitor(
    builtSchema,
    { directives, target: 'general', isTimestampFieldsAdded: true },
    { generate: CodeGenGenerateEnum.code },
  );
  visit(ast, { leave: visitor });
  visitor.generate();
  return visitor;
};

describe('AppSyncModelVisitor', () => {

  it('should support schema with id', () => {
    const schema = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: String!
        content: String!
      }
    `;
    const ast = parse(schema);
    const builtSchema = buildSchemaWithDirectives(schema);
    const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
    visit(ast, { leave: visitor });
    expect(visitor.models.Post).toBeDefined();

    const postFields = visitor.models.Post.fields;
    expect(postFields[0].name).toEqual('id');
    expect(postFields[0].type).toEqual('ID');
    expect(postFields[0].isNullable).toEqual(false);
    expect(postFields[0].isList).toEqual(false);
  });

  it('should throw error when schema has id of Non ID type', () => {
    const schema = /* GraphQL */ `
      type Post @model {
        id: String!
        title: String!
        content: String!
      }
    `;
    const ast = parse(schema);
    const builtSchema = buildSchemaWithDirectives(schema);
    const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
    expect(() => visit(ast, { leave: visitor })).toThrowError();
  });
  it('should have id as the first field to ensure arity of constructors', () => {
    const schema = /* GraphQL */ `
      type Post @model {
        title: String!
        content: String!
        id: ID!
      }
    `;
    const ast = parse(schema);
    const builtSchema = buildSchemaWithDirectives(schema);
    const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
    visit(ast, { leave: visitor });
    const postFields = visitor.models.Post.fields;
    expect(postFields[0].name).toEqual('id');
    expect(postFields[0].type).toEqual('ID');
    expect(postFields[0].isNullable).toEqual(false);
    expect(postFields[0].isList).toEqual(false);
  });

  it('should generate different version if field required attribute is different', () => {
    const schemaNotRequired = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: String
      }
    `;
    const visitorNotRequired = createAndGenerateVisitor(schemaNotRequired);

    const schemaRequired = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: String!
      }
    `;
    const visitorRequired = createAndGenerateVisitor(schemaRequired);

    const notRequiredVersion = (visitorNotRequired as any).computeVersion();
    const requiredVersion = (visitorRequired as any).computeVersion();

    expect(notRequiredVersion).not.toBe(requiredVersion);
  });

  it('should generate different version if field array attribute is different', () => {
    const schemaNoArray = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: String
      }
    `;
    const visitorNoArray = createAndGenerateVisitor(schemaNoArray);

    const schemaWithArray = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: [String]
      }
    `;
    const visitorWithArray = createAndGenerateVisitor(schemaWithArray);

    const noArrayVersion = (visitorNoArray as any).computeVersion();
    const withArrayVersion = (visitorWithArray as any).computeVersion();

    expect(noArrayVersion).not.toBe(withArrayVersion);
  });

  it('should generate different version if field array has required items', () => {
    const schemaNotRequired = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: [String]
      }
    `;
    const visitorNotRequired = createAndGenerateVisitor(schemaNotRequired);

    const schemaRequired = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: [String!]
      }
    `;
    const visitorRequired = createAndGenerateVisitor(schemaRequired);

    const notRequiredVersion = (visitorNotRequired as any).computeVersion();
    const requiredVersion = (visitorRequired as any).computeVersion();

    expect(notRequiredVersion).not.toBe(requiredVersion);
  });

  it('should generate different version if field array changes to required', () => {
    const schemaNotRequired = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: [String]
      }
    `;
    const visitorNotRequired = createAndGenerateVisitor(schemaNotRequired);

    const schemaRequired = /* GraphQL */ `
      type Post @model {
        id: ID!
        title: [String]!
      }
    `;
    const visitorRequired = createAndGenerateVisitor(schemaRequired);

    const notRequiredVersion = (visitorNotRequired as any).computeVersion();
    const requiredVersion = (visitorRequired as any).computeVersion();

    expect(notRequiredVersion).not.toBe(requiredVersion);
  });

  describe(' 2 Way Connection', () => {
    describe('with connection name', () => {
      const schema = /* GraphQL */ `
        type Post @model {
          id: ID!,
          title: String!
          content: String
          comments: [Comment] @connection(name: "PostComment")
        }

        type Comment @model {
          id: ID!,
          comment: String!
          post: Post @connection(name: "PostComment")
        }
      `;
      it('one to many connection', () => {
        const ast = parse(schema);
        const builtSchema = buildSchemaWithDirectives(schema);
        const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
        visit(ast, { leave: visitor });
        visitor.generate();
        const commentsField = visitor.models.Post.fields.find(f => f.name === 'comments');
        const postField = visitor.models.Comment.fields.find(f => f.name === 'post');
        expect(commentsField).toBeDefined();
        expect(commentsField!.connectionInfo).toBeDefined();
        const connectionInfo = (commentsField!.connectionInfo as any) as CodeGenFieldConnectionHasMany;
        expect(connectionInfo.kind).toEqual(CodeGenConnectionType.HAS_MANY);
        expect(connectionInfo.associatedWith).toEqual(postField);
        expect(connectionInfo.connectedModel).toEqual(visitor.models.Comment);
      });

      it('many to one connection', () => {
        const ast = parse(schema);
        const builtSchema = buildSchemaWithDirectives(schema);
        const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
        visit(ast, { leave: visitor });
        visitor.generate();
        const commentsField = visitor.models.Post.fields.find(f => f.name === 'comments');
        const postField = visitor.models.Comment.fields.find(f => f.name === 'post');
        expect(postField).toBeDefined();
        expect(postField!.connectionInfo).toBeDefined();
        const connectionInfo = (postField!.connectionInfo as any) as CodeGenFieldConnectionBelongsTo;
        expect(connectionInfo.kind).toEqual(CodeGenConnectionType.BELONGS_TO);
        expect(connectionInfo.connectedModel).toEqual(visitor.models.Post);
      });
    });
    describe('connection with fields argument', () => {
      const schema = /* GraphQL */ `
        type Post @model {
          id: ID!,
          title: String!
          content: String
          comments: [Comment] @connection(fields: ["id"])
        }

        type Comment @model {
          id: ID!,
          comment: String!
          postId: ID!
          post: Post @connection(fields: ["postId"])
        }
      `;

      it('one to many connection', () => {
        const ast = parse(schema);
        const builtSchema = buildSchemaWithDirectives(schema);
        const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
        visit(ast, { leave: visitor });
        visitor.generate();
        const commentsField = visitor.models.Post.fields.find(f => f.name === 'comments');
        const commentIdField = visitor.models.Post.fields.find(f => f.name === 'id');
        const postField = visitor.models.Comment.fields.find(f => f.name === 'post');
        expect(commentsField).toBeDefined();
        expect(commentsField!.connectionInfo).toBeDefined();
        const connectionInfo = (commentsField!.connectionInfo as any) as CodeGenFieldConnectionHasMany;
        expect(connectionInfo.kind).toEqual(CodeGenConnectionType.HAS_MANY);
        expect(connectionInfo.associatedWith).toEqual(commentIdField);
        expect(connectionInfo.connectedModel).toEqual(visitor.models.Comment);
      });

      it('many to one connection', () => {
        const ast = parse(schema);
        const builtSchema = buildSchemaWithDirectives(schema);
        const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
        visit(ast, { leave: visitor });
        visitor.generate();

        const postField = visitor.models.Comment.fields.find(f => f.name === 'post');
        expect(postField).toBeDefined();
        expect(postField!.connectionInfo).toBeDefined();
        const connectionInfo = (postField!.connectionInfo as any) as CodeGenFieldConnectionBelongsTo;
        expect(connectionInfo.kind).toEqual(CodeGenConnectionType.BELONGS_TO);
        expect(connectionInfo.connectedModel).toEqual(visitor.models.Post);
      });
    });
  });

  describe('one way connection', () => {
    it('should not include a comments in Post when comments field does not have connection directive', () => {
      const schema = /* GraphQL */ `
        type Post @model {
          id: ID!,
          title: String!
          content: String
          comments: [Comment]
        }

        type Comment @model {
          id: ID!,
          comment: String!
          post: Post @connection
        }
      `;
      const ast = parse(schema);
      const builtSchema = buildSchemaWithDirectives(schema);
      const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
      visit(ast, { leave: visitor });
      visitor.generate();
      const postFields = visitor.models.Post.fields.map(field => field.name);
      expect(postFields).not.toContain('comments');
    });

    it('should not include a post when post field in Comment when post does not have connection directive', () => {
      const schema = /* GraphQL */ `
        type Post @model {
          id: ID!,
          title: String!
          content: String
          comments: [Comment] @connection
        }

        type Comment @model {
          id: ID!,
          comment: String!
          post: Post
        }
      `;
      const ast = parse(schema);
      const builtSchema = buildSchemaWithDirectives(schema);
      const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
      visit(ast, { leave: visitor });
      visitor.generate();
      const commentsField = visitor.models.Comment.fields.map(field => field.name);
      expect(commentsField).not.toContain('post');
      expect(commentsField).toContain('postCommentsId'); // because of connection from Post.comments
    });
  });
  describe('auth directive', () => {
    it('should process auth with owner authorization', () => {
      const schema = /* GraphQL */ `
        type Post @searchable @model @auth(rules: [{ allow: owner }]) {
          id: ID!,
          title: String!
          content: String
        }
      `;
      const ast = parse(schema);
      const builtSchema = buildSchemaWithDirectives(schema);
      const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
      visit(ast, { leave: visitor });
      visitor.generate();
      const postModel = visitor.models.Post;
      const authDirective = postModel.directives.find(d => d.name === 'auth');
      expect(authDirective).toBeDefined();
      const authRules = authDirective!.arguments.rules;
      expect(authRules).toBeDefined();
      expect(authRules).toHaveLength(1);
      const ownerRule = authRules[0];
      expect(ownerRule).toBeDefined();

      expect(ownerRule.provider).toEqual('userPools');
      expect(ownerRule.identityClaim).toEqual('cognito:username');
      expect(ownerRule.ownerField).toEqual('owner');
      expect(ownerRule.operations).toEqual(['create', 'update', 'delete', 'read']);
    });

    it('should process group with owner authorization', () => {
      const schema = /* GraphQL */ `
        type Post @model @searchable @auth(rules: [{ allow: groups, groups: ["admin", "moderator"] }]) {
          id: ID!,
          title: String!
          content: String
        }
      `;
      const ast = parse(schema);
      const builtSchema = buildSchemaWithDirectives(schema);
      const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
      visit(ast, { leave: visitor });
      visitor.generate();
      const postModel = visitor.models.Post;
      const authDirective = postModel.directives.find(d => d.name === 'auth');
      expect(authDirective).toBeDefined();
      const authRules = authDirective!.arguments.rules;
      expect(authRules).toBeDefined();
      expect(authRules).toHaveLength(1);
      const groupRule = authRules[0];
      expect(groupRule).toBeDefined();
      expect(groupRule.provider).toEqual('userPools');
      expect(groupRule.groupClaim).toEqual('cognito:groups');
      expect(groupRule.operations).toEqual(['create', 'update', 'delete', 'read']);
    });

    it('should process field level auth with default owner authorization', () => {
      const schema = /* GraphQL*/ `
        type Employee @model
          @auth(rules: [
              { allow: owner },
              { allow: groups, groups: ["Admins"] }
          ]) {
            id: ID!
            name: String!
            address: String!
            ssn: String @auth(rules: [{allow: owner}])
        }
      `;
      const ast = parse(schema);
      const builtSchema = buildSchemaWithDirectives(schema);
      const visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
      visit(ast, { leave: visitor });
      visitor.generate();

      const ssnField = visitor.models.Employee.fields.find(f => f.name === 'ssn');
      const authDirective = ssnField?.directives.find(d => d.name === 'auth');
      expect(authDirective).toBeDefined();
      const authRules = authDirective!.arguments.rules;
      expect(authRules).toBeDefined();
      expect(authRules).toHaveLength(1);
      const ownerRule = authRules[0];
      expect(ownerRule).toBeDefined();

      expect(ownerRule.provider).toEqual('userPools');
      expect(ownerRule.identityClaim).toEqual('cognito:username');
      expect(ownerRule.ownerField).toEqual('owner');
      expect(ownerRule.operations).toEqual(['create', 'update', 'delete', 'read']);
    });
  });
  describe('model less type', () => {
    let visitor: AppSyncModelVisitor;
    beforeEach(() => {
      const schema = /* GraphQL */ `
        type Metadata {
          authorName: String!
          tags: [String]
          rating: Int!
        }
      `;
      const ast = parse(schema);
      const builtSchema = buildSchemaWithDirectives(schema);
      visitor = new AppSyncModelVisitor(builtSchema, { directives, target: 'android', generate: CodeGenGenerateEnum.code }, {});
      visit(ast, { leave: visitor });
      visitor.generate();
    });

    it('should support types without model', () => {
      const metadataType = visitor.nonModels.Metadata;

      expect(metadataType).toBeDefined();
      const metadataFields = metadataType.fields;

      expect(metadataFields[0].name).toEqual('authorName');
      expect(metadataFields[0].type).toEqual('String');
      expect(metadataFields[0].isNullable).toEqual(false);
      expect(metadataFields[0].isList).toEqual(false);

      expect(metadataFields[1].name).toEqual('tags');
      expect(metadataFields[1].type).toEqual('String');
      expect(metadataFields[1].isNullable).toEqual(true);
      expect(metadataFields[1].isList).toEqual(true);

      expect(metadataFields[2].name).toEqual('rating');
      expect(metadataFields[2].type).toEqual('Int');
      expect(metadataFields[2].isNullable).toEqual(false);
      expect(metadataFields[2].isList).toEqual(false);
    });
  });
  describe('timestamps for model directive', () => {
    it('should add timestamps fields in implicit cases', () => {
      const schema = /* GraphQL */ `
        type Post @model {
          id: ID!
        }
      `;
      const createdAtFieldObj = {
        name: 'createdAt',
        type: 'AWSDateTime',
        isList: false,
        isNullable: true,
        isReadOnly: true,
      };
      const updatedAtFieldObj = {
        name: 'updatedAt',
        type: 'AWSDateTime',
        isList: false,
        isNullable: true,
        isReadOnly: true,
      };
      const visitor = createAndGenerateVisitor(schema);
      expect(visitor.models.Post).toBeDefined();

      const postFields = visitor.models.Post.fields;
      const createdAtField = postFields.find(field => field.name === 'createdAt');
      expect(createdAtField).toMatchObject(createdAtFieldObj);
      const updatedAtField = postFields.find(field => field.name === 'updatedAt');
      expect(updatedAtField).toMatchObject(updatedAtFieldObj);
    });
    it('should add user defined timestamp fields in model directives', () => {
      const schema = /* GraphQL */ `
        type Post @model(timestamps: { createdAt: "createdOn", updatedAt: "updatedOn" }) {
          id: ID!
        }
      `;
      const createdAtFieldObj = {
        name: 'createdOn',
        type: 'AWSDateTime',
        isList: false,
        isNullable: true,
        isReadOnly: true,
      };
      const updatedAtFieldObj = {
        name: 'updatedOn',
        type: 'AWSDateTime',
        isList: false,
        isNullable: true,
        isReadOnly: true,
      };
      const visitor = createAndGenerateVisitor(schema);
      expect(visitor.models.Post).toBeDefined();

      const postFields = visitor.models.Post.fields;
      const createdAtField = postFields.find(field => field.name === 'createdOn');
      expect(createdAtField).toMatchObject(createdAtFieldObj);
      const updatedAtField = postFields.find(field => field.name === 'updatedOn');
      expect(updatedAtField).toMatchObject(updatedAtFieldObj);
    });
    it('should not override original fields if user define them explicitly in schema', () => {
      const schema = /* GraphQL */ `
        type Post @model {
          id: ID!
          createdAt: AWSDateTime!
          updatedAt: AWSDateTime!
        }
      `;
      const createdAtFieldObj = {
        name: 'createdAt',
        type: 'AWSDateTime',
        isList: false,
        isNullable: false,
      };
      const updatedAtFieldObj = {
        name: 'updatedAt',
        type: 'AWSDateTime',
        isList: false,
        isNullable: false,
      };
      const visitor = createAndGenerateVisitor(schema);
      expect(visitor.models.Post).toBeDefined();

      const postFields = visitor.models.Post.fields;
      const createdAtField = postFields.find(field => field.name === 'createdAt');
      expect(createdAtField).toMatchObject(createdAtFieldObj);
      expect(createdAtField.isReadOnly).not.toBeDefined();
      const updatedAtField = postFields.find(field => field.name === 'updatedAt');
      expect(updatedAtField).toMatchObject(updatedAtFieldObj);
    });
    it('should not override original fields if users define them explicitly in schema and use timestamps params in @model', () => {
      const schema = /* GraphQL */ `
        type Post @model(timestamps: { createdAt: "createdOn", updatedAt: "updatedOn" }) {
          id: ID!
          createdOn: AWSDateTime!
          updatedOn: AWSDateTime!
        }
      `;
      const createdAtFieldObj = {
        name: 'createdOn',
        type: 'AWSDateTime',
        isList: false,
        isNullable: false,
      };
      const updatedAtFieldObj = {
        name: 'updatedOn',
        type: 'AWSDateTime',
        isList: false,
        isNullable: false,
      };
      const visitor = createAndGenerateVisitor(schema);
      expect(visitor.models.Post).toBeDefined();

      const postFields = visitor.models.Post.fields;
      const createdAtField = postFields.find(field => field.name === 'createdOn');
      expect(createdAtField).toMatchObject(createdAtFieldObj);
      expect(createdAtField.isReadOnly).not.toBeDefined();
      const updatedAtField = postFields.find(field => field.name === 'updatedOn');
      expect(updatedAtField).toMatchObject(updatedAtFieldObj);
    });
  });
});
