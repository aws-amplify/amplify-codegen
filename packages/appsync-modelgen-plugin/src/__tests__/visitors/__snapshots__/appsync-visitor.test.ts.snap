// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSyncModelVisitor Other GraphQL types shoud support query, mutation and subscription types 1`] = `
Object {
  "echo": Object {
    "baseType": "String",
    "directives": Array [],
    "isList": false,
    "isNullable": true,
    "name": "echo",
    "operationType": "query",
    "parameters": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "msg",
        "type": "String",
      },
    ],
    "type": "String",
  },
  "echo2": Object {
    "baseType": "Todo",
    "directives": Array [],
    "isList": false,
    "isNullable": true,
    "name": "echo2",
    "operationType": "query",
    "parameters": Array [
      Object {
        "baseType": "ID",
        "directives": Array [],
        "isList": false,
        "isNullable": false,
        "name": "todoId",
        "type": "ID",
      },
    ],
    "type": "Todo",
  },
  "echo3": Object {
    "baseType": "Todo",
    "directives": Array [],
    "isList": true,
    "isListNullable": false,
    "isNullable": false,
    "name": "echo3",
    "operationType": "query",
    "parameters": Array [],
    "type": "Todo",
  },
  "echo4": Object {
    "baseType": "Phone",
    "directives": Array [],
    "isList": false,
    "isNullable": true,
    "name": "echo4",
    "operationType": "query",
    "parameters": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "number",
        "type": "String",
      },
    ],
    "type": "Phone",
  },
  "echo5": Object {
    "baseType": "CustomUnion",
    "directives": Array [],
    "isList": true,
    "isListNullable": false,
    "isNullable": false,
    "name": "echo5",
    "operationType": "query",
    "parameters": Array [],
    "type": "CustomUnion",
  },
  "echo6": Object {
    "baseType": "String",
    "directives": Array [],
    "isList": false,
    "isNullable": false,
    "name": "echo6",
    "operationType": "query",
    "parameters": Array [
      Object {
        "baseType": "CustomInput",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "customInput",
        "type": "CustomInput",
      },
    ],
    "type": "String",
  },
  "echo7": Object {
    "baseType": "ICustom",
    "directives": Array [],
    "isList": true,
    "isListNullable": false,
    "isNullable": true,
    "name": "echo7",
    "operationType": "query",
    "parameters": Array [],
    "type": "ICustom",
  },
  "getAllTodo": Object {
    "baseType": "String",
    "directives": Array [],
    "isList": false,
    "isNullable": true,
    "name": "getAllTodo",
    "operationType": "query",
    "parameters": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "msg",
        "type": "String",
      },
      Object {
        "baseType": "CustomInput",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "input",
        "type": "CustomInput",
      },
    ],
    "type": "String",
  },
}
`;

exports[`AppSyncModelVisitor Other GraphQL types shoud support query, mutation and subscription types 2`] = `
Object {
  "mutate": Object {
    "baseType": "Todo",
    "directives": Array [],
    "isList": false,
    "isNullable": true,
    "name": "mutate",
    "operationType": "mutation",
    "parameters": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": true,
        "isListNullable": false,
        "isNullable": false,
        "name": "msg",
        "type": "String",
      },
    ],
    "type": "Todo",
  },
}
`;

exports[`AppSyncModelVisitor Other GraphQL types shoud support query, mutation and subscription types 3`] = `
Object {
  "onMutate": Object {
    "baseType": "Todo",
    "directives": Array [],
    "isList": true,
    "isListNullable": true,
    "isNullable": false,
    "name": "onMutate",
    "operationType": "subscription",
    "parameters": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "msg",
        "type": "String",
      },
    ],
    "type": "Todo",
  },
}
`;

exports[`AppSyncModelVisitor Other GraphQL types should support input types 1`] = `
Object {
  "CustomInput": Object {
    "inputValues": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": false,
        "name": "customField1",
        "type": "String",
      },
      Object {
        "baseType": "BillingSource",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "customField2",
        "type": "BillingSource",
      },
      Object {
        "baseType": "NestedInput",
        "directives": Array [],
        "isList": false,
        "isNullable": false,
        "name": "customField3",
        "type": "NestedInput",
      },
    ],
    "name": "CustomInput",
    "type": "input",
  },
  "NestedInput": Object {
    "inputValues": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": false,
        "name": "content",
        "type": "String",
      },
    ],
    "name": "NestedInput",
    "type": "input",
  },
}
`;

exports[`AppSyncModelVisitor Other GraphQL types should support interface types 1`] = `
Object {
  "ICustom": Object {
    "fields": Array [
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": false,
        "name": "firstName",
        "parameters": Array [],
        "type": "String",
      },
      Object {
        "baseType": "String",
        "directives": Array [],
        "isList": false,
        "isNullable": true,
        "name": "lastName",
        "parameters": Array [],
        "type": "String",
      },
      Object {
        "baseType": "INestedCustom",
        "directives": Array [],
        "isList": true,
        "isListNullable": false,
        "isNullable": false,
        "name": "birthdays",
        "parameters": Array [],
        "type": "INestedCustom",
      },
    ],
    "name": "ICustom",
    "type": "interface",
  },
  "INestedCustom": Object {
    "fields": Array [
      Object {
        "baseType": "AWSDate",
        "directives": Array [],
        "isList": false,
        "isNullable": false,
        "name": "birthDay",
        "parameters": Array [],
        "type": "AWSDate",
      },
    ],
    "name": "INestedCustom",
    "type": "interface",
  },
}
`;

exports[`AppSyncModelVisitor Other GraphQL types should support union types 1`] = `
Object {
  "CustomUnion": Object {
    "name": "CustomUnion",
    "type": "union",
    "typeNames": Array [
      "Todo",
      "Phone",
    ],
  },
}
`;
